# HIV Long Carriage Case Report


[![DOI](https://zenodo.org/badge/464699387.svg)](https://zenodo.org/badge/latestdoi/464699387)


Amplicon genomes were analysed using the SIGNAL workflow v1.4.4dev (#60dd466) 9
(https://github.com/jaleezyy/covid-19-signal) and ncov-tools v1.8.0 (https://github.com/jts/ncov-
tools) with the ARTICv3 primer schemes and the MN908947.3 reference genome. Annotated
variant calling data and all additional analysis code, including calculation of non-
synonymous/synonymous changes per site (Ka/Ks) can be found here.

## Files

    HIV_mut_gain_loss.ipynb : analysis notebook
    long_carriage_genomes_lineages.csv : pangolin (pangoLEARN) lineage assignments
    20210830_MN01513_SRI_deepseq_summary_valid_negative_pass_only.tsv : ncov-tools QC from SIGNAL output
    genomes/ : consensus sequences from SIGNAL
    variants/ : variant calling files from SIGNAL annotated with ncov-tools (SnpEff)
    figures/ : Figures generated by notebook
    table/ : Tables generated by notebook
