# HIV Long Carriage Case Report


[![DOI](https://zenodo.org/badge/464699387.svg)](https://zenodo.org/badge/latestdoi/464699387)


Amplicon genomes were analysed using the SIGNAL workflow v1.4.4dev (#60dd466) 9
(https://github.com/jaleezyy/covid-19-signal) and ncov-tools v1.8.0 (https://github.com/jts/ncov-
tools) with the ARTICv3 primer schemes and the MN908947.3 reference genome. Annotated
variant calling data and all additional analysis code, including calculation of non-
synonymous/synonymous changes per site (Ka/Ks) can be found here.

## Methods
Testing for SARS-CoV-2 was performed on nasopharyngeal swabs by extraction and reverse-transcription polymerase chain reaction (RT-PCR) targeting two SARS-CoV-2 regions, the 5’ untranslated region (UTR) and the envelope (E) gene. Cycle threshold (Ct) values were determined using Rotor-Gene Q software (Qiagen, https://www.qiagen.com), with Ct values of <35 considered positive, and those with detectable RNA but Ct values of 35 or a single target detected considered indeterminate.8

Serum was tested for the presence of antibodies against SARS-CoV-2 using the Roche Elecsys anti-SARS-CoV-2 immunoassays. The assays detect total antibody levels for antibodies against the spike and nucleocapsid antigens, with anti-spike values of <0.8U/mL considered negative and anti-nucleocapsid cutoff index of <1.0 considered non-reactive.

Vero E6 cells were inoculated in duplicate with 250 µL of sample collected from the patient, as previously described.8 Cells were monitored for up to six days post inoculation for evidence of cytopathic effect (CPE). RNA was extracted from culture supernatants using the QIAmp Viral RNA Mini kit and analyzed by RT-PCR for SARS-CoV-2 UTR and E genes. Ct values were compared to those of the original nasopharyngeal swab used to infect cells. CPE-positive cultures with Ct values lower than the original swab indicate presence of actively replicating virus.

Viral titers were determined for nasopharyngeal swabs and isolated cultures. Briefly, neat or tenfold serially diluted samples were added to Vero E6 cells in three or six replicates and incubated for up to 5-6 days at 37℃, 5% CO2 before the 50% tissue culture infectious dose (TCID50) was measured. The Spearman and Karber method was used to calculate TCID50/mL.

Amplicon genomes were analysed using the SIGNAL workflow v1.4.4dev (#60dd466)9 (https://github.com/jaleezyy/covid-19-signal) and ncov-tools v1.8.0 (https://github.com/jts/ncov-tools) with the ARTICv3 primer schemes and the MN908947.3 reference genome. Annotated variant calling data and all additional analysis code, including calculation of genome-wide non-synonymous/synonymous changes per site (Ka/Ks) relative to "USA/CA-CDC-STM-000008272" can be found in this repository.

## Files
    HIV_mut_gain_loss.ipynb : analysis notebook
    long_carriage_genomes_lineages.csv : pangolin (pangoLEARN) lineage assignments
    20210830_MN01513_SRI_deepseq_summary_valid_negative_pass_only.tsv : ncov-tools QC from SIGNAL output
    genomes/ : consensus sequences from SIGNAL
    variants/ : variant calling files from SIGNAL annotated with ncov-tools (SnpEff)
    figures/ : Figures generated by notebook
    table/ : Tables generated by notebook
    
## References
    Banerjee A, Nasir JA, Budylowski P, et al. Isolation, Sequence, Infectivity, and Replication Kinetics of Severe Acute Respiratory Syndrome Coronavirus 2. Emerg Infect Dis 2020; 26:2054–63.
    
    Nasir JA, Kozak RA, Aftanas P, et al. A Comparison of Whole Genome Sequencing of SARS-CoV-2 Using Amplicon-Based Sequencing, Random Hexamers, and Bait Capture. Viruses 2020; 12:895. 

